extern/test_bmi_cpp/src/test_bmi_cpp.cpp

```diff
@@ -0,0 +1,429 @@
+#include "test_bmi_cpp.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+
+std::string TestBmiCpp::GetComponentName(){
+  return "Testing BMI C++ Model";
+}
+
+double TestBmiCpp::GetCurrentTime(){
+  return current_model_time;
+}
+
+double TestBmiCpp::GetEndTime(){
+  double time = this->GetStartTime();
+  time += this->num_time_steps * this->time_step_size;
+  return time;
+}
+
+int TestBmiCpp::GetGridRank(const int grid){
+  if (grid == 0) {
+      return 1;
+  }
+  else {
+    throw new std::runtime_error("Rank requested for non-existent grid.");
```

Comment:
@hellkite500, does that make this function better or worse for testing purposes?  I.e., is this what we are likely to see in the wild?

Reply:
I think he's right on this one...I'm not sure why I did this, as on review it does seem like a bad idea and generally frowned upon. I think I was having trouble getting reference types to match up and hit on the wrong answer.

Reply:
I agree that isn't great code according to typical criteria, but not-great code is a reality.  I'm really asking whether our testing coverage needs code like this in the test library, because we expect the framework to be used with real-world modules that do something like this.

Reply:
I think we should make a complete test example the best way we can...this will be the example people can use to model their own integrations with.  If we want a unit test that tests the handling of a dynamically allocated exception, then we should mock that in isoloation.

Reply:
Yes, besides strictly being a test "mock" these classes are used as examples or even boilerplate to show users how to implement... so I'm not in favor of putting undesirable things in here just to test.